quic/s2n-quic-core/src/inet/ipv4.rs

```diff
@@ -305,6 +314,330 @@ impl From<IpV4Address> for [u8; IPV4_LEN] {
+            .field(
+                "flags (more fragments)",
+                &self.flag_fragment.more_fragments(),
+            )
+            .field("fragment_offset", &self.flag_fragment.fragment_offset())
+            .field("ttl", &self.ttl)
+            .field("protocol", &self.protocol)
+            .field("checksum", &format_args!("0x{:04x}", self.checksum.get()))
+            .field("source", &self.source)
+            .field("destination", &self.destination)
+            .finish()
+    }
+}
+
+impl Header {
+    /// Swaps the direction of the header
+    #[inline]
+    pub fn swap(&mut self) {
+        core::mem::swap(&mut self.source, &mut self.destination)
+    }
+
+    #[inline]
+    pub const fn vihl(&self) -> &Vihl {
+        &self.vihl
+    }
+
+    #[inline]
+    pub fn vihl_mut(&mut self) -> &mut Vihl {
+        &mut self.vihl
+    }
+
+    #[inline]
+    pub const fn tos(&self) -> &Tos {
+        &self.tos
+    }
+
+    #[inline]
+    pub fn tos_mut(&mut self) -> &mut Tos {
+        &mut self.tos
+    }
+
+    #[inline]
+    pub const fn total_len(&self) -> &U16 {
+        &self.total_len
+    }
+
+    #[inline]
+    pub fn total_len_mut(&mut self) -> &mut U16 {
+        &mut self.total_len
+    }
+
+    #[inline]
+    pub const fn id(&self) -> &U16 {
+        &self.id
+    }
+
+    #[inline]
+    pub fn id_mut(&mut self) -> &mut U16 {
+        &mut self.id
+    }
+
+    #[inline]
+    pub const fn flag_fragment(&self) -> &FlagFragment {
+        &self.flag_fragment
+    }
+
+    #[inline]
+    pub fn flag_fragment_mut(&mut self) -> &mut FlagFragment {
+        &mut self.flag_fragment
+    }
+
+    #[inline]
+    pub const fn ttl(&self) -> &u8 {
+        &self.ttl
+    }
+
+    #[inline]
+    pub fn ttl_mut(&mut self) -> &mut u8 {
+        &mut self.ttl
+    }
+
+    #[inline]
+    pub const fn protocol(&self) -> &ip::Protocol {
+        &self.protocol
+    }
+
+    #[inline]
+    pub fn protocol_mut(&mut self) -> &mut ip::Protocol {
+        &mut self.protocol
+    }
+
+    #[inline]
+    pub const fn checksum(&self) -> &U16 {
+        &self.checksum
+    }
+
+    #[inline]
+    pub fn checksum_mut(&mut self) -> &mut U16 {
+        &mut self.checksum
+    }
+
+    #[inline]
+    pub const fn source(&self) -> &IpV4Address {
+        &self.source
+    }
+
+    #[inline]
+    pub fn source_mut(&mut self) -> &mut IpV4Address {
+        &mut self.source
+    }
+
+    #[inline]
+    pub const fn destination(&self) -> &IpV4Address {
+        &self.destination
+    }
+
+    #[inline]
+    pub fn destination_mut(&mut self) -> &mut IpV4Address {
+        &mut self.destination
+    }
+}
+
+define_inet_type!(
+    pub struct Vihl {
+        value: u8,
+    }
+);
+
+impl fmt::Debug for Vihl {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("Vihl")
+            .field("version", &self.version())
+            .field("header_len", &self.header_len())
+            .finish()
+    }
+}
+
+impl Vihl {
```

Comment:
It would be nice to explicitly call out any opaque boundaries. E.g. `Vihl` represents `Version` information in the most significant 4 bits and `IHL` is the least significant 4 bits. Although this is pretty easily inferred from the RFC diagram, it's nice to give people help when trying to decipher bit operations.

Reply:
Similar thing for the `Tos` struct.

Reply:
I can add some notes